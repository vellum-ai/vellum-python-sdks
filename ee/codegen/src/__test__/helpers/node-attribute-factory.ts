import { NodeAttribute } from "src/types/vellum";

export const nodeAttributeFactory = (
  id: string,
  name: string,
  attributes: NodeAttribute[]
): NodeAttribute => ({
  id,
  name,
  value: {
    type: "CONSTANT_VALUE",
    value: {
      type: "JSON",
      value: attributes,
    },
  },
});
