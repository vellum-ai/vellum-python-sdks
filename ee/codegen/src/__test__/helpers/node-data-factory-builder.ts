import {
  AdornmentNode,
  CodeResourceDefinition,
  NodeAttribute,
  NodeOutput,
  NodePort,
  NodeTrigger,
  WorkflowDataNode as WorkflowDataNodeType,
} from "src/types/vellum";

export class NodeDataFactoryBuilder<T extends WorkflowDataNodeType> {
  public nodeData: T;

  constructor(nodeData: T) {
    this.nodeData = nodeData;
  }

  withPorts(ports: NodePort[]): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      ports: ports,
    };
    return this;
  }

  withAttributes(attributes: NodeAttribute[]): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      attributes,
    };
    return this;
  }

  withAdornments(adornments: AdornmentNode[]): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      adornments,
    };
    return this;
  }

  withOutputs(outputs: NodeOutput[]): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      outputs,
    };
    return this;
  }

  withBase(base: CodeResourceDefinition): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      base,
    };
    return this;
  }

  withTrigger(trigger: NodeTrigger): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      trigger,
    };
    return this;
  }

  build(): T {
    return this.nodeData;
  }
}
