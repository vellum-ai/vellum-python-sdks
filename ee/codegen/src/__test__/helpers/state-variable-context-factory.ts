import { VellumVariable } from "vellum-ai/api/types";

import { WorkflowContext } from "src/context";
import { StateVariableContext } from "src/context/state-variable-context";

export function stateVariableContextFactory({
  stateVariableData,
  workflowContext,
}: {
  stateVariableData: VellumVariable;
  workflowContext: WorkflowContext;
}): StateVariableContext {
  return new StateVariableContext({
    stateVariableData,
    workflowContext,
  });
}
