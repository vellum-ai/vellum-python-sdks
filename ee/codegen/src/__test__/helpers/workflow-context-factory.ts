import { entrypointNodeDataFactory } from "./node-data-factories";

import { WorkflowContext } from "src/context";

export function workflowContextFactory({
  absolutePathToOutputDirectory,
  moduleName,
  workflowClassName,
  workflowRawData,
  strict = true,
  disableFormatting = true,
}: Partial<WorkflowContext.Args> = {}): WorkflowContext {
  return new WorkflowContext({
    absolutePathToOutputDirectory:
      absolutePathToOutputDirectory || "./src/__tests__/",
    moduleName: moduleName || "code",
    workflowClassName: workflowClassName || "TestWorkflow",
    vellumApiKey: "<TEST_API_KEY>",
    workflowRawData: workflowRawData || {
      nodes: [entrypointNodeDataFactory()],
      edges: [],
    },
    strict,
    disableFormatting,
  });
}
