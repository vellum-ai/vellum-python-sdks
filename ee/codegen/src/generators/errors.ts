export type CodegenErrorCode =
  | "PROJECT_SERIALIZATION_ERROR"
  | "WORKFLOW_GENERATION_ERROR"
  | "WORKFLOW_INPUT_GENERATION_ERROR"
  | "NODE_DEFINITION_GENERATION_ERROR"
  | "NODE_ATTRIBUTE_GENERATION_ERROR"
  | "NODE_PORT_GENERATION_ERROR"
  | "NODE_NOT_FOUND_ERROR"
  | "NODE_OUTPUT_NOT_FOUND_ERROR"
  | "NODE_INPUT_NOT_FOUND_ERROR"
  | "UNSUPPORTED_SANDBOX_INPUT_ERROR"
  | "ENTITY_NOT_FOUND_ERROR"
  | "POST_PROCESSING_ERROR"
  | "VALUE_GENERATION_ERROR";

export abstract class BaseCodegenError extends Error {
  abstract code: CodegenErrorCode;
}

/**
 * An error that raises when deserializing the request to codegen
 * into a valid Workflow project.
 */
export class ProjectSerializationError extends BaseCodegenError {
  code = "PROJECT_SERIALIZATION_ERROR" as const;
}

/**
 * An error that raises when the Workflow definition fails to
 * generate.
 */
export class WorkflowGenerationError extends BaseCodegenError {
  code = "WORKFLOW_GENERATION_ERROR" as const;
}

/**
 * An error that raises when the Workflow Inpus fail to
 * generate.
 */
export class WorkflowInputGenerationError extends BaseCodegenError {
  code = "WORKFLOW_INPUT_GENERATION_ERROR" as const;
}

/**
 * An error that raises when the Node definition fails to generate
 */
export class NodeDefinitionGenerationError extends BaseCodegenError {
  code = "NODE_DEFINITION_GENERATION_ERROR" as const;
}

/**
 * An error that raises when generating a node attribute fails.
 */
export class NodeAttributeGenerationError extends BaseCodegenError {
  code = "NODE_ATTRIBUTE_GENERATION_ERROR" as const;
}

/**
 * An error that raises when generating a node port fails.
 */
export class NodePortGenerationError extends BaseCodegenError {
  code = "NODE_PORT_GENERATION_ERROR" as const;
}

/**
 * An error that raises when a node is not found.
 */
export class NodeNotFoundError extends BaseCodegenError {
  code = "NODE_NOT_FOUND_ERROR" as const;
}

/**
 * An error that raises when a vellum entity is not found.
 */
export class EntityNotFoundError extends BaseCodegenError {
  code = "ENTITY_NOT_FOUND_ERROR" as const;
}

/**
 * An error that raises when a sandbox input is not supported.
 */
export class UnsupportedSandboxInputError extends BaseCodegenError {
  code = "UNSUPPORTED_SANDBOX_INPUT_ERROR" as const;
}

/**
 * An error that raises when a node output is not found.
 */
export class NodeOutputNotFoundError extends BaseCodegenError {
  code = "NODE_OUTPUT_NOT_FOUND_ERROR" as const;
}

/**
 * An error that raises when a node input is not found.
 */
export class NodeInputNotFoundError extends BaseCodegenError {
  code = "NODE_INPUT_NOT_FOUND_ERROR" as const;
}

/**
 * An error that raises when the code fails to post process
 */
export class PostProcessingError extends BaseCodegenError {
  code = "POST_PROCESSING_ERROR" as const;
}

/**
 * An error that raises when a value fails to generate
 */
export class ValueGenerationError extends BaseCodegenError {
  code = "VALUE_GENERATION_ERROR" as const;
}
