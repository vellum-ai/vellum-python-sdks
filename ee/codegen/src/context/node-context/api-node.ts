import { WorkspaceSecrets as WorkspaceSecretsClient } from "vellum-ai/api/resources/workspaceSecrets/client/Client";
import { VellumError } from "vellum-ai/errors";

import { BaseNodeContext } from "src/context/node-context/base";
import { PortContext } from "src/context/port-context";
import { EntityNotFoundError } from "src/generators/errors";
import {
  ApiNode as ApiNodeType,
  NodeInput,
  WorkspaceSecretPointer,
} from "src/types/vellum";

export class ApiNodeContext extends BaseNodeContext<ApiNodeType> {
  baseNodeClassName = "APINode";
  baseNodeDisplayClassName = "BaseAPINodeDisplay";

  getNodeOutputNamesById(): Record<string, string> {
    return {
      [this.nodeData.data.jsonOutputId]: "json",
      [this.nodeData.data.statusCodeOutputId]: "status_code",
      [this.nodeData.data.textOutputId]: "text",
      ...(this.nodeData.data.errorOutputId
        ? { [this.nodeData.data.errorOutputId]: "error" }
        : {}),
    };
  }

  protected createPortContexts(): PortContext[] {
    return [
      new PortContext({
        workflowContext: this.workflowContext,
        nodeContext: this,
        portId: this.nodeData.data.sourceHandleId,
      }),
    ];
  }

  private async processSecretInput(input: NodeInput): Promise<void> {
    const inputRule = input?.value.rules.find(
      (rule): rule is WorkspaceSecretPointer => rule.type == "WORKSPACE_SECRET"
    );
    if (!inputRule || !inputRule.data?.workspaceSecretId) {
      return;
    }
    try {
      const tokenItem = await new WorkspaceSecretsClient({
        apiKey: this.workflowContext.vellumApiKey,
      }).retrieve(inputRule.data.workspaceSecretId);
      inputRule.data.workspaceSecretId = tokenItem.name;
    } catch (e) {
      if (e instanceof VellumError && e.statusCode === 404) {
        this.workflowContext.addError(
          new EntityNotFoundError(
            `Workspace Secret for attribute "${input.key}" not found.`,
            "WARNING"
          )
        );
      } else {
        throw e;
      }
    }
  }

  async buildProperties(): Promise<void> {
    const apiKeyInputId = this.nodeData.data.apiKeyHeaderValueInputId;
    const apiKeyInput = this.nodeData.inputs.find(
      (input) => input.id === apiKeyInputId
    );

    const bearerKeyInputId = this.nodeData.data.bearerTokenValueInputId;
    const bearerKeyInput = this.nodeData.inputs.find(
      (input) => input.id === bearerKeyInputId
    );
    const secrets = [];
    if (apiKeyInput) {
      secrets.push(apiKeyInput);
    }
    if (bearerKeyInput) {
      secrets.push(bearerKeyInput);
    }

    this.nodeData.data.additionalHeaders?.forEach((header) => {
      const headerInputValue = this.nodeData.inputs.find(
        (input) => input.id === header.headerValueInputId
      );
      if (headerInputValue) {
        secrets.push(headerInputValue);
      }
    });
    await Promise.all(
      secrets.map(async (input) => this.processSecretInput(input))
    );
  }
}
