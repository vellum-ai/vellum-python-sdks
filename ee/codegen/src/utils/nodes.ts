import { VellumError } from "vellum-ai/errors";

import { WorkflowNode } from "src/types/vellum";

export function getNodeLabel(nodeData: WorkflowNode): string {
  switch (nodeData.type) {
    case "GENERIC":
      return nodeData.definition?.name ?? nodeData.label ?? "Generic Node";
    default:
      return nodeData.data.label;
  }
}

export function isUnaryOperator(operator: string): boolean {
  return operator === "null" || operator === "notNull";
}

export function isVellumErrorWithDetail(
  error: unknown
): error is VellumError & { body: { detail: string } } {
  return (
    error instanceof VellumError &&
    typeof error.body === "object" &&
    error.body !== null &&
    "detail" in error.body &&
    typeof error.body.detail === "string"
  );
}
