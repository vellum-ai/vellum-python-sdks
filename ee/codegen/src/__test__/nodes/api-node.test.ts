import { Writer } from "@fern-api/python-ast/core/Writer";
import { v4 as uuidv4 } from "uuid";
import { SecretTypeEnum, WorkspaceSecretRead } from "vellum-ai/api";
import { WorkspaceSecrets } from "vellum-ai/api/resources/workspaceSecrets/client/Client";
import { VellumError } from "vellum-ai/errors/VellumError";
import { beforeEach, describe } from "vitest";

import { workflowContextFactory } from "src/__test__/helpers";
import { inputVariableContextFactory } from "src/__test__/helpers/input-variable-context-factory";
import {
  apiNodeFactory,
  ApiNodeFactoryProps,
  nodeInputFactory,
} from "src/__test__/helpers/node-data-factories";
import { createNodeContext, WorkflowContext } from "src/context";
import { ApiNodeContext } from "src/context/node-context/api-node";
import { EntityNotFoundError } from "src/generators/errors";
import { ApiNode } from "src/generators/nodes/api-node";

describe("ApiNode", () => {
  let workflowContext: WorkflowContext;
  let writer: Writer;
  let node: ApiNode;

  beforeEach(() => {
    workflowContext = workflowContextFactory();
    writer = new Writer();

    workflowContext.addInputVariableContext(
      inputVariableContextFactory({
        inputVariableData: {
          id: "5f64210f-ec43-48ce-ae40-40a9ba4c4c11",
          key: "additional_header_value",
          type: "STRING",
        },
        workflowContext,
      })
    );
    workflowContext.addInputVariableContext(
      inputVariableContextFactory({
        inputVariableData: {
          id: "b81c5c88-9528-47d0-8106-14a75520ed47",
          key: "additional_header_value",
          type: "STRING",
        },
        workflowContext,
      })
    );
  });

  const mockWorkspaceSecretDefinition = (workspaceSecret: {
    id: string;
    name: string;
  }): WorkspaceSecretRead => ({
    id: workspaceSecret.id,
    name: workspaceSecret.name,
    modified: new Date(),
    label: "mocked-workspace-secret-label",
    secretType: SecretTypeEnum.UserDefined,
  });

  const createNode = async ({
    workspaceSecrets,
    ...apiNodeProps
  }: {
    workspaceSecrets: { id: string; name: string }[];
  } & ApiNodeFactoryProps) => {
    const workspaceSecretIdByName = Object.fromEntries(
      workspaceSecrets.map(({ id, name }) => [name, id])
    );
    const workspaceSecretNameById = Object.fromEntries(
      workspaceSecrets.map(({ id, name }) => [id, name])
    );
    vi.spyOn(WorkspaceSecrets.prototype, "retrieve").mockImplementation(
      async (idOrName) => {
        const id = workspaceSecretIdByName[idOrName];
        if (id) {
          return mockWorkspaceSecretDefinition({
            id,
            name: idOrName,
          });
        }

        const name = workspaceSecretNameById[idOrName];
        if (name) {
          return mockWorkspaceSecretDefinition({
            id: idOrName,
            name,
          });
        }

        throw new Error(`Workspace secret ${idOrName} not found`);
      }
    );

    const nodeData = apiNodeFactory(apiNodeProps);

    const nodeContext = (await createNodeContext({
      workflowContext,
      nodeData,
    })) as ApiNodeContext;

    return new ApiNode({
      workflowContext,
      nodeContext,
    });
  };

  describe("basic auth secret node", () => {
    it.each([{ id: "1234", name: "test-secret" }])(
      "secret ids should show names",
      async (workspaceSecret: { id: string; name: string }) => {
        node = await createNode({
          workspaceSecrets: [workspaceSecret],
          bearerToken: nodeInputFactory({
            key: "bearer_token_value",
            value: {
              type: "WORKSPACE_SECRET",
              data: {
                type: "STRING",
                workspaceSecretId: workspaceSecret.id,
              },
            },
          }),
          apiKeyHeaderValue: nodeInputFactory({
            key: "api_key_header_value",
            value: {
              type: "WORKSPACE_SECRET",
              data: {
                type: "STRING",
                workspaceSecretId: workspaceSecret.id,
              },
            },
          }),
        });
        node.getNodeFile().write(writer);
        expect(await writer.toStringFormatted()).toMatchSnapshot();
      }
    );

    it("should generate Workspace secrets for header values", async () => {
      const workspaceSecretId = uuidv4();
      node = await createNode({
        workspaceSecrets: [{ id: workspaceSecretId, name: "test-secret" }],
        additionalHeaders: [
          {
            key: nodeInputFactory({
              key: "test-key",
              value: {
                type: "CONSTANT_VALUE",
                data: {
                  type: "STRING",
                  value: "X-Secret-Key",
                },
              },
            }),
            value: nodeInputFactory({
              key: "test-value",
              value: {
                type: "WORKSPACE_SECRET",
                data: {
                  type: "STRING",
                  workspaceSecretId: workspaceSecretId,
                },
              },
            }),
          },
        ],
      });
      node.getNodeFile().write(writer);
      expect(await writer.toStringFormatted()).toMatchSnapshot();
    });
  });

  describe("basic", () => {
    beforeEach(async () => {
      const nodeData = apiNodeFactory();

      const nodeContext = (await createNodeContext({
        workflowContext,
        nodeData,
      })) as ApiNodeContext;

      node = new ApiNode({
        workflowContext: workflowContext,
        nodeContext,
      });
    });

    it("getNodeFile", async () => {
      node.getNodeFile().write(writer);
      expect(await writer.toStringFormatted()).toMatchSnapshot();
    });

    it("getNodeDisplayFile", async () => {
      node.getNodeDisplayFile().write(writer);
      expect(await writer.toStringFormatted()).toMatchSnapshot();
    });
  });

  describe("reject on error enabled", () => {
    beforeEach(async () => {
      const nodeData = apiNodeFactory({
        errorOutputId: "af589f73-effe-4a80-b48f-fb912ac6ce67",
      });

      const nodeContext = (await createNodeContext({
        workflowContext,
        nodeData,
      })) as ApiNodeContext;

      node = new ApiNode({
        workflowContext: workflowContext,
        nodeContext,
      });
    });

    it("getNodeFile", async () => {
      node.getNodeFile().write(writer);
      expect(await writer.toStringFormatted()).toMatchSnapshot();
    });

    it("getNodeDisplayFile", async () => {
      node.getNodeDisplayFile().write(writer);
      expect(await writer.toStringFormatted()).toMatchSnapshot();
    });
  });

  describe("skip authorization type input id field if undefined", () => {
    beforeEach(async () => {
      const nodeData = apiNodeFactory({ authorizationTypeInput: null });

      const nodeContext = (await createNodeContext({
        workflowContext,
        nodeData,
      })) as ApiNodeContext;

      node = new ApiNode({
        workflowContext: workflowContext,
        nodeContext,
      });
    });

    it("getNodeFile", async () => {
      node.getNodeFile().write(writer);
      expect(await writer.toStringFormatted()).toMatchSnapshot();
    });
  });

  describe("skip bearer token if secret not found", () => {
    it("getNodeFile", async () => {
      vi.spyOn(WorkspaceSecrets.prototype, "retrieve").mockImplementation(
        async (idOrName) => {
          throw new VellumError({
            message: `Workspace secret '${idOrName}' not found`,
            statusCode: 404,
          });
        }
      );
      const nodeData = apiNodeFactory({
        bearerToken: nodeInputFactory({
          key: "bearer_token_value",
          value: {
            type: "WORKSPACE_SECRET",
            data: {
              type: "STRING",
              workspaceSecretId: "some-non-existent-workspace-secret-id",
            },
          },
        }),
      });

      const workflowContext = workflowContextFactory({ strict: false });
      const nodeContext = (await createNodeContext({
        workflowContext,
        nodeData,
      })) as ApiNodeContext;

      node = new ApiNode({
        workflowContext: workflowContext,
        nodeContext,
      });

      node.getNodeFile().write(writer);
      expect(await writer.toStringFormatted()).toMatchSnapshot();
      expect(workflowContext.getErrors()).toHaveLength(1);
      const error = workflowContext.getErrors()[0];
      expect(error).toBeInstanceOf(EntityNotFoundError);
      expect(error?.message).toContain(
        'Workspace Secret for attribute "bearer_token_value" not found.'
      );
      expect(error?.severity).toBe("WARNING");
    });
  });
});
