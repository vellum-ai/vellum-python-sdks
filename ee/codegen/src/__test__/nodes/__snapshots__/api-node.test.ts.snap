// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ApiNode > basic > getNodeDisplayFile 1`] = `
"from uuid import UUID

from vellum_ee.workflows.display.nodes import BaseAPINodeDisplay
from vellum_ee.workflows.display.nodes.types import (
    NodeOutputDisplay,
    PortDisplayOverrides,
)
from vellum_ee.workflows.display.vellum import NodeDisplayData, NodeDisplayPosition

from ...nodes.api_node import APINode


class APINodeDisplay(BaseAPINodeDisplay[APINode]):
    label = "API Node"
    node_id = UUID("2cd960a3-cb8a-43ed-9e3f-f003fc480951")
    target_handle_id = UUID("06573a05-e6f0-48b9-bc6e-07e06d0bc1b1")
    url_input_id = UUID("480a4c12-22d6-4223-a38a-85db5eda118c")
    method_input_id = UUID("9bf086d4-feed-47ff-9736-a5a6aa3a11cc")
    body_input_id = UUID("74865eb7-cdaf-4d40-a499-0a6505e72680")
    authorization_type_input_id = UUID("de330dac-05b1-4e78-bee7-7452203af3d5")
    bearer_token_value_input_id = UUID("931502c1-23a5-4e2a-a75e-80736c42f3c9")
    api_key_header_key_input_id = UUID("96c8343d-cc94-4df0-9001-eb2905a00be7")
    api_key_header_value_input_id = UUID("bfc2e790-66fd-42fd-acf7-3b2c785c1a0a")
    additional_header_key_input_ids = {
        "foo": UUID("8ad006f3-d19e-4af1-931f-3e955152cd91"),
        "bar": UUID("3075be8c-248b-421d-9266-7779297be883"),
        "baz": UUID("13c2dd5e-cdd0-431b-aa91-46ad8da1cb51"),
    }
    additional_header_value_input_ids = {
        "foo": UUID("36865dca-40b4-432c-bab4-1e11bb9f4083"),
        "bar": UUID("00baaee1-b785-403d-b391-f68b3aea334f"),
        "baz": UUID("408c2b3d-7c30-4e01-a2e3-276753beadbc"),
    }
    node_input_ids_by_name = {
        "method": UUID("9bf086d4-feed-47ff-9736-a5a6aa3a11cc"),
        "url": UUID("480a4c12-22d6-4223-a38a-85db5eda118c"),
        "body": UUID("74865eb7-cdaf-4d40-a499-0a6505e72680"),
        "authorization_type": UUID("de330dac-05b1-4e78-bee7-7452203af3d5"),
        "bearer_token_value": UUID("931502c1-23a5-4e2a-a75e-80736c42f3c9"),
        "api_key_header_key": UUID("96c8343d-cc94-4df0-9001-eb2905a00be7"),
        "api_key_header_value": UUID("bfc2e790-66fd-42fd-acf7-3b2c785c1a0a"),
        "additional_header_key_1": UUID("8ad006f3-d19e-4af1-931f-3e955152cd91"),
        "additional_header_value_1": UUID("36865dca-40b4-432c-bab4-1e11bb9f4083"),
        "additional_header_key_2": UUID("3075be8c-248b-421d-9266-7779297be883"),
        "additional_header_value_2": UUID("00baaee1-b785-403d-b391-f68b3aea334f"),
        "additional_header_key_3": UUID("13c2dd5e-cdd0-431b-aa91-46ad8da1cb51"),
        "additional_header_value_3": UUID("408c2b3d-7c30-4e01-a2e3-276753beadbc"),
    }
    output_display = {
        APINode.Outputs.json: NodeOutputDisplay(
            id=UUID("af576eaa-d39d-4c19-8992-1f01a65a709a"), name="json"
        ),
        APINode.Outputs.status_code: NodeOutputDisplay(
            id=UUID("69250713-617d-42a4-9326-456c70d0ef20"), name="status_code"
        ),
        APINode.Outputs.text: NodeOutputDisplay(
            id=UUID("81b270c0-4deb-4db3-aae5-138f79531b2b"), name="text"
        ),
    }
    port_displays = {
        APINode.Ports.default: PortDisplayOverrides(
            id=UUID("c38a71f6-3ffb-45fa-9eea-93c6984a9e3e")
        )
    }
    display_data = NodeDisplayData(
        position=NodeDisplayPosition(x=2075.7067885117494, y=234.65663468515768),
        width=462,
        height=288,
    )
"
`;

exports[`ApiNode > basic > getNodeFile 1`] = `
"from vellum.workflows.constants import APIRequestMethod, AuthorizationType
from vellum.workflows.nodes.displayable import APINode as BaseAPINode


class APINode(BaseAPINode):
    url = "https://example.vellum.ai"
    method = APIRequestMethod.POST
    headers = {
        "foo": "foo-value",
        "bar": "bar-value",
        "baz": "baz-value",
    }
    api_key_header_key = None
    authorization_type = AuthorizationType.API_KEY
    api_key_header_value = "<my-api-value>"
    bearer_token_value = "<my-bearer-token>"
"
`;

exports[`ApiNode > basic api node > should generate API key header fields when API key is provided 1`] = `
"from vellum.workflows.constants import AuthorizationType
from vellum.workflows.nodes.displayable import APINode as BaseAPINode


class APINode(BaseAPINode):
    url = "https://example.vellum.ai"
    api_key_header_key = "X-API-KEY"
    authorization_type = AuthorizationType.API_KEY
    api_key_header_value = "my-api-key"
"
`;

exports[`ApiNode > basic api node > should generate all fields when all conditions are provided 1`] = `
"from vellum.workflows.constants import APIRequestMethod, AuthorizationType
from vellum.workflows.nodes.displayable import APINode as BaseAPINode


class APINode(BaseAPINode):
    url = "https://example.vellum.ai"
    method = APIRequestMethod.POST
    json = {
        "foo": "bar",
    }
    headers = {
        "Content-Type": "application/json",
    }
    api_key_header_key = "X-API-KEY"
    authorization_type = AuthorizationType.API_KEY
    api_key_header_value = "my-api-key"
"
`;

exports[`ApiNode > basic api node > should generate headers field when headers are provided 1`] = `
"from vellum.workflows.nodes.displayable import APINode as BaseAPINode


class APINode(BaseAPINode):
    url = "https://example.vellum.ai"
    headers = {
        "Content-Type": "application/json",
    }
"
`;

exports[`ApiNode > basic api node > should generate method field when method is POST 1`] = `
"from vellum.workflows.constants import APIRequestMethod
from vellum.workflows.nodes.displayable import APINode as BaseAPINode


class APINode(BaseAPINode):
    url = "https://example.vellum.ai"
    method = APIRequestMethod.POST
"
`;

exports[`ApiNode > basic api node > should not generate body field when body is empty 1`] = `
"from vellum.workflows.nodes.displayable import APINode as BaseAPINode


class APINode(BaseAPINode):
    url = "https://example.vellum.ai"
"
`;

exports[`ApiNode > basic api node > should only generate url field 1`] = `
"from vellum.workflows.nodes.displayable import APINode as BaseAPINode


class APINode(BaseAPINode):
    url = "https://example.vellum.ai"
"
`;

exports[`ApiNode > basic auth secret node > secret ids should show names 1`] = `
"from vellum.workflows.constants import APIRequestMethod, AuthorizationType
from vellum.workflows.nodes.displayable import APINode as BaseAPINode
from vellum.workflows.references import VellumSecretReference


class APINode(BaseAPINode):
    url = "https://example.vellum.ai"
    method = APIRequestMethod.POST
    headers = {
        "foo": "foo-value",
        "bar": "bar-value",
        "baz": "baz-value",
    }
    api_key_header_key = None
    authorization_type = AuthorizationType.API_KEY
    api_key_header_value = VellumSecretReference("test-secret")
    bearer_token_value = VellumSecretReference("test-secret")
"
`;

exports[`ApiNode > basic auth secret node > should generate Workspace secrets for header values 1`] = `
"from vellum.workflows.constants import APIRequestMethod, AuthorizationType
from vellum.workflows.nodes.displayable import APINode as BaseAPINode
from vellum.workflows.references import VellumSecretReference


class APINode(BaseAPINode):
    url = "https://example.vellum.ai"
    method = APIRequestMethod.POST
    headers = {
        "X-Secret-Key": VellumSecretReference("test-secret"),
    }
    api_key_header_key = None
    authorization_type = AuthorizationType.API_KEY
    api_key_header_value = "<my-api-value>"
    bearer_token_value = "<my-bearer-token>"
"
`;

exports[`ApiNode > reject on error enabled > getNodeDisplayFile 1`] = `
"from uuid import UUID

from vellum_ee.workflows.display.nodes import BaseAPINodeDisplay, BaseTryNodeDisplay
from vellum_ee.workflows.display.nodes.types import (
    NodeOutputDisplay,
    PortDisplayOverrides,
)
from vellum_ee.workflows.display.vellum import NodeDisplayData, NodeDisplayPosition

from ...nodes.api_node import APINode


@BaseTryNodeDisplay.wrap(error_output_id=UUID("af589f73-effe-4a80-b48f-fb912ac6ce67"))
class APINodeDisplay(BaseAPINodeDisplay[APINode]):
    label = "API Node"
    node_id = UUID("2cd960a3-cb8a-43ed-9e3f-f003fc480951")
    target_handle_id = UUID("06573a05-e6f0-48b9-bc6e-07e06d0bc1b1")
    url_input_id = UUID("480a4c12-22d6-4223-a38a-85db5eda118c")
    method_input_id = UUID("9bf086d4-feed-47ff-9736-a5a6aa3a11cc")
    body_input_id = UUID("74865eb7-cdaf-4d40-a499-0a6505e72680")
    authorization_type_input_id = UUID("de330dac-05b1-4e78-bee7-7452203af3d5")
    bearer_token_value_input_id = UUID("931502c1-23a5-4e2a-a75e-80736c42f3c9")
    api_key_header_key_input_id = UUID("96c8343d-cc94-4df0-9001-eb2905a00be7")
    api_key_header_value_input_id = UUID("bfc2e790-66fd-42fd-acf7-3b2c785c1a0a")
    additional_header_key_input_ids = {
        "foo": UUID("8ad006f3-d19e-4af1-931f-3e955152cd91"),
        "bar": UUID("3075be8c-248b-421d-9266-7779297be883"),
        "baz": UUID("13c2dd5e-cdd0-431b-aa91-46ad8da1cb51"),
    }
    additional_header_value_input_ids = {
        "foo": UUID("36865dca-40b4-432c-bab4-1e11bb9f4083"),
        "bar": UUID("00baaee1-b785-403d-b391-f68b3aea334f"),
        "baz": UUID("408c2b3d-7c30-4e01-a2e3-276753beadbc"),
    }
    node_input_ids_by_name = {
        "method": UUID("9bf086d4-feed-47ff-9736-a5a6aa3a11cc"),
        "url": UUID("480a4c12-22d6-4223-a38a-85db5eda118c"),
        "body": UUID("74865eb7-cdaf-4d40-a499-0a6505e72680"),
        "authorization_type": UUID("de330dac-05b1-4e78-bee7-7452203af3d5"),
        "bearer_token_value": UUID("931502c1-23a5-4e2a-a75e-80736c42f3c9"),
        "api_key_header_key": UUID("96c8343d-cc94-4df0-9001-eb2905a00be7"),
        "api_key_header_value": UUID("bfc2e790-66fd-42fd-acf7-3b2c785c1a0a"),
        "additional_header_key_1": UUID("8ad006f3-d19e-4af1-931f-3e955152cd91"),
        "additional_header_value_1": UUID("36865dca-40b4-432c-bab4-1e11bb9f4083"),
        "additional_header_key_2": UUID("3075be8c-248b-421d-9266-7779297be883"),
        "additional_header_value_2": UUID("00baaee1-b785-403d-b391-f68b3aea334f"),
        "additional_header_key_3": UUID("13c2dd5e-cdd0-431b-aa91-46ad8da1cb51"),
        "additional_header_value_3": UUID("408c2b3d-7c30-4e01-a2e3-276753beadbc"),
    }
    output_display = {
        APINode.Outputs.json: NodeOutputDisplay(
            id=UUID("af576eaa-d39d-4c19-8992-1f01a65a709a"), name="json"
        ),
        APINode.Outputs.status_code: NodeOutputDisplay(
            id=UUID("69250713-617d-42a4-9326-456c70d0ef20"), name="status_code"
        ),
        APINode.Outputs.text: NodeOutputDisplay(
            id=UUID("81b270c0-4deb-4db3-aae5-138f79531b2b"), name="text"
        ),
    }
    port_displays = {
        APINode.Ports.default: PortDisplayOverrides(
            id=UUID("c38a71f6-3ffb-45fa-9eea-93c6984a9e3e")
        )
    }
    display_data = NodeDisplayData(
        position=NodeDisplayPosition(x=2075.7067885117494, y=234.65663468515768),
        width=462,
        height=288,
    )
"
`;

exports[`ApiNode > reject on error enabled > getNodeFile 1`] = `
"from vellum.workflows.constants import APIRequestMethod, AuthorizationType
from vellum.workflows.nodes.core import TryNode
from vellum.workflows.nodes.displayable import APINode as BaseAPINode


@TryNode.wrap()
class APINode(BaseAPINode):
    url = "https://example.vellum.ai"
    method = APIRequestMethod.POST
    headers = {
        "foo": "foo-value",
        "bar": "bar-value",
        "baz": "baz-value",
    }
    api_key_header_key = None
    authorization_type = AuthorizationType.API_KEY
    api_key_header_value = "<my-api-value>"
    bearer_token_value = "<my-bearer-token>"
"
`;

exports[`ApiNode > skip authorization type input id field if undefined > getNodeFile 1`] = `
"from vellum.workflows.constants import APIRequestMethod
from vellum.workflows.nodes.displayable import APINode as BaseAPINode


class APINode(BaseAPINode):
    url = "https://example.vellum.ai"
    method = APIRequestMethod.POST
    headers = {
        "foo": "foo-value",
        "bar": "bar-value",
        "baz": "baz-value",
    }
    api_key_header_key = None
"
`;

exports[`ApiNode > skip bearer token if secret not found > getNodeFile 1`] = `
"from vellum.workflows.constants import APIRequestMethod, AuthorizationType
from vellum.workflows.nodes.displayable import APINode as BaseAPINode
from vellum.workflows.references import VellumSecretReference


class APINode(BaseAPINode):
    url = "https://example.vellum.ai"
    method = APIRequestMethod.POST
    headers = {
        "foo": "foo-value",
        "bar": "bar-value",
        "baz": "baz-value",
    }
    api_key_header_key = None
    authorization_type = AuthorizationType.API_KEY
    api_key_header_value = "<my-api-value>"
    bearer_token_value = VellumSecretReference("some-non-existent-workspace-secret-id")
"
`;
